Make multilocale libraries determine their port information on the fly

lydia-duncan opened the pull request:
This change modifies the generation of multilocale libraries to use
ephemeral ports instead of hard-coded ones when communicating
between the client and spawned server.  It does this via a secret
flag that is not visible to the user, `--chpl-mli-socket-loc`, which
communicates the connection information where the client will
listen.  The server will then create sockets for communicating the
function calls, arguments, and return values of the exported
functions, and will use the client's connection socket to
communicate the connection information for those sockets.

Passed testing over std/ paratest with futures and a full gasnet paratest.

dlongnecke-cray commented:
It's strange to me that we have to use a flag to sneak our information through the launcher. In normal execution of Chapel multi-locale libraries, the end user has to be able to pass through an unadorned argv/argc to the entrypoint, no?

What is preventing us from just leveraging that? 

Just curious to know a bit more about how the launcher whines at us.

lydia-duncan commented:
Basically, what was going on was that `parseArgs` was traversing the arguments, looking for configs and other known flags.  This was happening before `chpl_library_init` was called on the server side.  I am appending the connection information to the end of the already present argv/argc information, but 1) parseArgs didn't know anything about whether we were getting called from a multilocale library or execution settting and 2) it seemed hooky to just "assume the last argument" was what we wanted given that situation.  Having an additional flag that parseArgs recognized and ignored allowed us to indicate to parseArgs that the situation was appropriate

dlongnecke-cray commented:
Ok, sounds reasonable. Thanks for clarifying. My only worry is making sure a user can't (whether accidentally or maliciously) get ahold of the flag somehow?

lydia-duncan commented:
That's a good point.  I should add some additional checking

dlongnecke-cray commented on runtime/etc/src/mli/client_runtime.c:
```diff
@@ -81,30 +95,50 @@ void chpl_library_init(int argc, char** argv) {
   chpl_mli_client_init(&chpl_client);
 
-  chpl_mli_bind(chpl_client.setup_sock, ip, setupPort);
+  chpl_mli_bind(chpl_client.setup_sock, ip, "*");
```
I think I wrote some logic that converts "localhost" -> "*" in `chpl_mli_bind`, not sure that's needed now if we're just going to pass an ephemeral device in directly.

lydia-duncan replied:
Yeah, I debated about updating bind as well as connect.  I think it would be reasonable to just pass in *:*, or even just call bind with only the socket as an argument, since we won't be using explicit port and hostname information for that call at all

dlongnecke-cray replied:
Yeah, I suppose some refactoring is in order for those functions.

dlongnecke-cray commented on runtime/etc/src/mli/client_runtime.c:
```diff
@@ -81,30 +95,50 @@ void chpl_library_init(int argc, char** argv) {
-  chpl_mli_client_launch(argc, argv);
-
+  // Send the setup socket as the last argument when launching the server.
+  int argcPlusSock = argc + 2;
```
It's not that I'm "dead set" on it, but...now that it was brought up earlier through #13030, I'd kind of like us to stick to a consistent code style across these modules.

I'm not sure how egregious things are already (I'll admit, I wasn't really thinking about consistency earlier and I defaulted to `snake_case`), but I'd be willing to settle on something (strange, around these parts...) like `snake_case_for_functions` and `camelCase` for variables.

Or does it not matter? I'll defer to you.

dlongnecke-cray commented on runtime/etc/src/mli/server_runtime.c:
```diff
@@ -55,42 +55,59 @@ void chpl_mli_server_deinit(struct chpl_mli_context* server) {
+static
+void chpl_mli_push_connection(char* connection) {
+  int len = strlen(connection) + 1;
+  chpl_mli_debugf("Pushing expected size %d\n", len);
```
Colon here?

dlongnecke-cray commented on runtime/etc/src/mli/client_runtime.c:
```diff
@@ -81,30 +95,50 @@ void chpl_library_init(int argc, char** argv) {
+  }
+  argvPlusSock[argc] = "--chpl-mli-socket-loc";
+  argvPlusSock[argc + 1] = setup_sock_conn;
+  chpl_mli_debugf("spawning server with %d args\n", argcPlusSock);
```
Capitalize "passing"? I don't know if we care about proper capitalization and stuff for debug messages or not, since realistically we're the only ones who are ever going to see them. 

I normally don't care too much about formatting, but #13030 has me up in arms...

dlongnecke-cray commented on runtime/etc/src/mli/server_runtime.c:
```diff
@@ -55,42 +55,59 @@ void chpl_mli_server_deinit(struct chpl_mli_context* server) {
+  chpl_mli_debugf("Main port on: %s\n", main_conn);
+
+  chpl_mli_bind(chpl_server.arg, ip, "*");
+  char* arg_conn = chpl_mli_connection_info(chpl_server.arg);
```
We could make a function for this chunk?

lydia-duncan replied:
Eh, I think it's short enough that I'd rather not.  We'd have to send in both the socket and a string to describe it in the debug message, and that seems like too much tooling for three lines.

dlongnecke-cray replied:
Sounds good.

dlongnecke-cray commented on runtime/src/main_mli.c:
```diff
@@ -24,18 +24,19 @@
+  // Drop down into the server listen loop, passing in the last argument, which
+  // will be the socket connection information for communicating where the
+  // server will be listening.
+  chpl_mli_smain(argv[argc-1]);
```
Hmmm, I'm thinking that we might want to refactor the listen loop into "setup" -> "listen" -> "teardown" functions. Not necessarily with this PR, though.

dlongnecke-cray replied:
It's good that we continue to keep the actual body of `main` as small as possible, though.

lydia-duncan commented:
I think I've addressed your feedback (except for the one I responded to directly just now).  I need to:
- check testing when I'm back in the office
- muse a little more about how to error when the secret flag is thrown for single-locale/non-library programs
- poke the weird connection issue for my home

lydia-duncan commented:
But I think it would be good for you to take a look at what's here now, and let me know if you think any of those three tasks shouldn't hold up the PR (currently leaning towards number 3 not holding it up)

lydia-duncan commented:
(current branch passed full paratest with futures for CHPL_COMM=none, so I didn't break that at least)

dlongnecke-cray commented:
It looks good, I like the error messages in `runtime/src/arg.c`. I don't think the coverage for the secret flag needs to be 100% complete to be able to merge this (if we waited till everything was 100% complete, we'd have nothing at all...).
